packages/flutter/lib/src/material/input_border.dart

```diff
@@ -116,6 +116,91 @@ class _NoInputBorder extends InputBorder {
   }
 }
 
+/// Clips an [InputDecorator]'s [InputDecoration.filled] background to a rounded
+/// rectangle.
+///
+/// This class is similar to [OutlineBorder] except that no outline is drawn.
```

Comment:
No outline is drawn and the shape is filled, right?

What if you want a fill _and_ an outline?

Reply:
(maybe the right solution is just to add the ability to control the fill color to OutlineBorder?)

Reply:
Honestly, it feels like all this should be one class with params.

Reply:
Well, except the layouts are different in the styles too.

Reply:
So, the `.filled` param works with `OutlineBorder`. This class came about because you can't set some params on the existing border classes. 

Expectation: Pass a `BorderSide` to `OutlineInputBorde`r that has a `width` of `0.0`.

Actual:` InputDecorator` has hardcoded border widths that override that `0.0` depending on state if you use the `border` convenience API.
Possible Solution: They could check for `borderWidth` being null or use the values as multipliers instead.

Expectation: Passing a `width` of `0.0` doesn't show on screen.

Actual: Lines with a `width` of `0.0` are drawn as a hairline so passing 
```
enabledBorder: OutlineInputBorder(
  borderSide: BorderSide.none
),
```
doesn't do exactly what I expected it to which is not draw.
>   /// Since a zero width is normally painted as a hairline width rather than no
  /// border at all, the zero factor is special-cased to instead change the
  /// style no [BorderStyle.none].
>

Possible Solution: Make hairline still easy but more explicit with a high-level function that returns the width in the points you need to get one pixel:
```
double SinglePixelInPoints(BuildContext context) {
  return 1 / MediaQuery.of(context).devicePixelRatio;
}
```

And a width of 0.0 draws nothing.

Reply:
Perhaps the special case that's needed here is OutlineBorder(borderSide: BorderSide.none).

Reply:
So, `OutlineInputBorder` does have `borderSide: BorderSide.none`:

```
const OutlineInputBorder({
    BorderSide borderSide = BorderSide.none,
    this.borderRadius = const BorderRadius.all(Radius.circular(4.0)),
    this.gapPadding = 4.0,
  }) : assert(borderRadius != null),
       assert(gapPadding != null && gapPadding >= 0.0),
       super(borderSide: borderSide);
```
And the InputDecorator turns that into `1.0` or `2.0` depending on state:

```
InputBorder _getDefaultBorder(ThemeData themeData) {
...
double borderWeight;
if (decoration.isCollapsed || decoration?.border == InputBorder.none || !decoration.enabled)
  borderWeight = 0.0;
else
  borderWeight = isFocused ? 2.0 : 1.0;

...
return border.copyWith(borderSide: BorderSide(color: borderColor, width: borderWeight));
```
I think the code should be changed to look for `null` instead of `.none` to add the default values.

Reply:
And of course it's possible to be explicit by state:
```
enabledBorder: OutlineInputBorder(
  borderSide: BorderSide.none
),
```